services/galley/src/Galley/API/Teams.hs

```diff
@@ -677,26 +693,32 @@ ensureNotElevated targetPermissions member =
     )
     $ throwM invalidPermissions
 
-addTeamMemberInternal :: TeamId -> Maybe UserId -> Maybe ConnId -> NewTeamMember -> [TeamMember] -> Galley ()
-addTeamMemberInternal tid origin originConn newMem mems = do
+ensureNotTooLarge :: TeamId -> Galley TeamSize
+ensureNotTooLarge tid = do
+  o <- view options
+  (TeamSize size) <- BrigTeam.getSize tid
+  unless (size < fromIntegral (o ^. optSettings . setMaxTeamSize)) $
+    throwM tooManyTeamMembers
+  return $ TeamSize size
+
+addTeamMemberInternal :: TeamId -> Maybe UserId -> Maybe ConnId -> NewTeamMember -> Data.TeamMemberList -> Galley TeamSize
+addTeamMemberInternal tid origin originConn newMem memList = do
   let new = newMem ^. ntmNewTeamMember
   Log.debug $
     Log.field "targets" (toByteString (new ^. userId))
       . Log.field "action" (Log.val "Teams.addTeamMemberInternal")
-  o <- view options
-  (TeamSize teamSize) <- BrigTeam.getSize tid
-  unless (teamSize < fromIntegral (o ^. optSettings . setMaxTeamSize)) $
-    throwM tooManyTeamMembers
+  sizeBeforeAdd <- ensureNotTooLarge tid
   Data.addTeamMember tid new
   cc <- filter (view managedConversation) <$> Data.teamConversations tid
   now <- liftIO getCurrentTime
   for_ cc $ \c ->
     Data.addMember now (c ^. conversationId) (new ^. userId)
   let e = newEvent MemberJoin tid now & eventData .~ Just (EdMemberJoin (new ^. userId))
-  push1 $ newPush1 (new ^. userId) (TeamEvent e) (recipients origin new) & pushConn .~ originConn
+  push1 $ newPush1 (Data.teamMemberListType memList) (new ^. userId) (TeamEvent e) (recipients origin new) & pushConn .~ originConn
+  return sizeBeforeAdd
   where
-    recipients (Just o) n = list1 (userRecipient o) (membersToRecipients (Just o) (n : mems))
-    recipients Nothing n = list1 (userRecipient (n ^. userId)) (membersToRecipients Nothing (n : mems))
+    recipients (Just o) n = list1 (userRecipient o) (membersToRecipients (Just o) (n : Data.teamMembers memList))
+    recipients Nothing n  = list1 (userRecipient (n ^. userId)) (membersToRecipients Nothing (Data.teamMembers memList))
```

Comment:
Note that we had some duplicate user here which was causing some of my tests to fail